Remove third-party notice for local commands

weierophinney opened the pull request:
This patch provides functionality to exclude local commands (defined as anything not in the vendor directory) from resulting in a prompt warning that a third-party command is about to be executed (when executed as part of a chain).

It does so by checking if the command class is defined in the configured Composer `vendor-dir`, but not within a laminas, laminas-api-tools, or mezzio subdirectory. If it is one of those subdirectories, or not defined in the `vendor-dir`, we consider it a local command.

Fixes #44

weierophinney committed:
feat: do not mark local commands as third-party commands

This patch provides functionality to exclude local commands (defined as anything not in the vendor directory) from resulting in a prompt warning that a third-party command is about to be executed (when executed as part of a chain).

It does so by by checking to see if it is defined in the configured vendor directory.
The logic excludes classes installed via Laminas, Laminas API Tools, or Mezzio packages.

Fixes #44

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
Adds CHANGELOG entry for #53

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

michalbundyra commented on src/Listener/TerminateListener.php:
```diff
@@ -165,4 +176,50 @@ private function validateInputMap(array $inputMap, string $commandClass): void
+        Assert::isMap($namespaceMap);
+
+        $localNamespaces = [];
+        $vendorDir       = $basePath . '/vendor';
```
It might be not always `vendor`, see:
https://getcomposer.org/doc/06-config.md#vendor-dir

weierophinney replied:
Ugh, I always forget that you _can_ do this. I'll update to resolve it.

I've been thinking about your idea, and I think it's likely the more correct path forward. The main problem I have is it's going to be next-to-impossible to test that it detects 3rd party commands correctly, unless we install some as part of our dev dependencies.

michalbundyra replied:
I guess we can mock it somehow, in the worst case scenario - create a file in tests in vendor 😱

weierophinney replied:
Found a way, using the "Show" command provided with Psalm. :smile:

michalbundyra reviewed (commented):
@weierophinney in general it's fine, but I had a thought about it, and maybe it would be better  (instead of checking the namespace) to check what is the location of the file?

So we would need to show the message if the source file is in vendor but not in:
- vendor/laminas/
- vendor/laminas-api-tools/
- vendor/mezzio/

I think it might be a bit better than checking namespace, because anyone can define any namespace, but I believe noone else is not able to register any `laminas/`, `mezzio/`, `laminas-api-tools/` package.
Also we do not need to care on all cases - PSR-0, PSR-4, files autoloading etc...

Side note: Of course if someone is using `Laminas\` namespace it's wrong, but we do not have too much control on it.

Thoughts?

weierophinney commented:
@michalbundyra Okay, updated the check for third-party commands such that this now looks to see if the command class is defined within the configured `vendor-dir`, and **is not** in one of the laminas, laminas-api-tools, or mezzio subdirectories.

weierophinney commented:
I've tested this in an app I maintain, and can confirm:

- I no longer get the "third party command" notice for app-specific commands.
- Chained commands continue to work correctly.
- All of the above is true even when `config.vendor-dir` is configured to another path.

michalbundyra commented on src/Listener/TerminateListener.php:
```diff
@@ -165,4 +179,58 @@ private function validateInputMap(array $inputMap, string $commandClass): void
+
+        $vendorDir = $composer['config']['vendor-dir'] ?? $basePath . '/vendor';
+        Assert::string($vendorDir);
+        Assert::directory($vendorDir);
```
this might be extreme case, but possible according to composer documentation:

> `$HOME` and `~` will be replaced by your home directory's path

https://getcomposer.org/doc/06-config.md#vendor-dir

michalbundyra commented on src/Listener/TerminateListener.php:
```diff
@@ -165,4 +179,58 @@ private function validateInputMap(array $inputMap, string $commandClass): void
+    /**
+     * @psalm-param class-string $class
+     */
+    private function matchesLocalNamespace(string $class, string $vendorDir): bool
```
I would suggest to change the method name, as we are not matching the vendor dir not the namespace

michalbundyra commented on src/Listener/TerminateListener.php:
```diff
@@ -165,4 +179,58 @@ private function validateInputMap(array $inputMap, string $commandClass): void
+        $composer = json_decode($composerJson, true);
+        Assert::isMap($composer);
+
+        $vendorDir = $composer['config']['vendor-dir'] ?? $basePath . '/vendor';
```
I cannot see any test for custom vendor-dir, so I guess you have checked it manually?

I had an impression first that it might not working first, as it can be relative, but as a fallback we are using fullpath, but noticed then couple lines below realpath on the result, so I believe it's fine :)

weierophinney replied:
Yeah, had to test it manually, since we resolve the `composer.json` based on the current working directory. I considered a few options for testing, but they all required a lot of bizarre hoops and coding remnants that would have only existed for testing.

When I manually tested, I did so by changing the value of `config.vendor-dir` in an existing project, and having it install my branch as the 0.1.6 release, and then checking how chained commands worked. Prior to the changes in this branch, I got the warning about using 3rd party code; after, the notices were gone.

weierophinney replied:
Ended up using reflection to test the method, as the change to resolve home paths meant I needed to test those code paths as well.

michalbundyra reviewed (approved):
@weierophinney Looks good, one comment to suggest method name change and one possible very edge case with custom vendor-dir. Your call if you want to handle these. Thanks

weierophinney committed:
feat: resolve "~" and "$HOME" to current home directory

This patch addresses a comment in the Composer documentation (https://getcomposer.org/doc/06-config.md#vendor-dir):

> `$HOME` and `~` will be replaced by your home directory's path

I've added unit tests for the `getVendorDirectory` method, using reflection to test it, which allows ensuring the directory is resolved per configuration.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
